cluster_pack/uploader.py

```diff
@@ -309,11 +309,25 @@ def _upload_env_from_venv(
         if not resolved_fs.exists(dir):
             resolved_fs.mkdir(dir)
         _logger.info(f'Uploading env at {local_package_path} to {package_path}')
-        resolved_fs.put(local_package_path, package_path)
+        _put(resolved_fs, local_package_path, package_path)
 
         _dump_archive_metadata(package_path, reqs, resolved_fs)
 
 
+def _put(resolved_fs: filesystem.EnhancedFileSystem,
+         source_path: str,
+         destination_path: str) -> None:
+    with tempfile.TemporaryDirectory() as temp_dir:
+        temp_subdir = os.path.basename(temp_dir)
+        upload_dir = os.path.join(os.path.dirname(destination_path), temp_subdir)
+        try:
+            resolved_fs.put(source_path, upload_dir)
+            resolved_fs.move(upload_dir, destination_path)

```

Comment:
Don't we still have a concurrency issue here if you have 2 calls trying to move it here ? Shouldn't we check if destination_path might not be already there as mentionned in the ticket: 

process 1: venv does not exist on hdfs at /final/venv, decide to upload to /tmp1/venv

process 2: venv does not exist on hdfs at /final/venv, decide to upload to /tmp2/venv

process 1: mv /tmp1/venv /final/venv

process 2: mv /tmp2/venv /final/venv

process 1 succeeds

process 2 fails and double checks that the existing pex is the same as the one he was uploading

If same -> good nothing to do

If not same -> raise

final state: /final/venv is present and contains the venv